### PR TITLE
[VPlan] Move recording of Inst->VPValue to VPRecipeBuilder (NFCI).

### PR DESCRIPTION
**Pull request by fhahn**
Instead of keeping a mapping of Inst->VPValues (of their corresponding recipes) in VPlan's Value2VPValue mapping, keep it in VPRecipeBuilder instead. After recently replacing the last user of this mapping after initial construction, this mapping is only needed for recipe construction (to map IR operands to VPValue operands).

By moving the mapping, VPlan's VPValue tracking can be simplified and limited only to live-ins. It also allows removing disableValue2VPValue and associated machinery & asserts.

### CODE REVIEW

**Comment by ayalz** [comment]
> Instead of keeping a mapping of Inst->VPValues (of their corresponding recipes) in VPlan's Value2VPValue mapping, keep it in VPRecipeBuilder instead. After recently replacing the last user of this mapping after initial construction, this mapping is only needed for recipe construction (to map IR operands to VPValue operands).
> 
> By moving the mapping, VPlan's VPValue tracking can be simplified and limited only to live-ins. It also allows removing disableValue2VPValue and associated machinery & asserts.

Nice clean-up! Conceptually analogous to TransformState which keeps track of VPValue-to-Value(s) mapping during codegen only. Independent of this patch: should VPRecipeBuilder (or some other sub/class thereof) be confined to initial VPlan construction from LLVM-IR, excluding construction of recipes during VPlan-to-VPlan transformations?

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 1]
```diff
@@ -117,13 +116,8 @@ class VPRecipeBuilder {
                                        VFRange &Range, VPBasicBlock *VPBB,
                                        VPlanPtr &Plan);
 
-  /// Set the recipe created for given ingredient. This operation is a no-op for
-  /// ingredients that were not marked using a nullptr entry in the map.
+  /// Set the recipe created for given ingredient.
   void setRecipe(Instruction *I, VPRecipeBase *R) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:120]: Added assert, thanks!

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 2]
```diff
@@ -172,6 +158,19 @@ class VPRecipeBuilder {
   /// Add the incoming values from the backedge to reduction & first-order
   /// recurrence cross-iteration phis.
   void fixHeaderPhis();
+
+  /// Returns a range mapping the values the range \p Operands to their
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h]: Done, thanks!

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 3]
```diff
@@ -172,6 +158,19 @@ class VPRecipeBuilder {
...
+  /// corresponding VPValues.
+  iterator_range<mapped_iterator<Use *, std::function<VPValue *(Value *)>>>
+  mapToVPValues(User::op_range Operands, VPlan &Plan);
+
+  VPValue *getVPValue(Value *V, VPlan &Plan) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h]: Updated, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 4]
```diff
@@ -2985,33 +2977,30 @@ class VPlan {
   void setName(const Twine &newName) { Name = newName.str(); }
 
   void addVPValue(Value *V, VPValue *VPV) {
-    assert((Value2VPValueEnabled || VPV->isLiveIn()) &&
-           "Value2VPValue mapping may be out of date!");
-    assert(V && "Trying to add a null Value to VPlan");
-    assert(!Value2VPValue.count(V) && "Value already exists in VPlan");
+    assert(VPV->isLiveIn());
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: added, thanks, also changed to addLiveIn

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 5]
```diff
@@ -2985,33 +2977,30 @@ class VPlan {
   void setName(const Twine &newName) { Name = newName.str(); }
 
   void addVPValue(Value *V, VPValue *VPV) {
-    assert((Value2VPValueEnabled || VPV->isLiveIn()) &&
-           "Value2VPValue mapping may be out of date!");
-    assert(V && "Trying to add a null Value to VPlan");
-    assert(!Value2VPValue.count(V) && "Value already exists in VPlan");
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: retained, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 6]
```diff
@@ -2985,33 +2977,30 @@ class VPlan {
...
-           "Value2VPValue mapping may be out of date!");
+    assert(Value2VPValue[V]->isLiveIn() &&
+           "Only live-ins should be in mapping");
     return Value2VPValue[V];
   }
 
   /// Gets the VPValue for \p V or adds a new live-in (if none exists yet) for
   /// \p V.
-  VPValue *getVPValueOrAddLiveIn(Value *V) {
+  VPValue *getOrAddLiveIn(Value *V) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: I think it is more accurate (and compact), but I removed that change, to reduce the diff

**Review by fhahn — COMMENTED** [review]
> Independent of this patch: should VPRecipeBuilder (or some other sub/class thereof) be confined to initial VPlan construction from LLVM-IR, excluding construction of recipes during VPlan-to-VPlan transformations?

Yes, `VPRecipeBuilder` ( not to be confused with `VPBuilder`) is only used for initial construction. We should split of the logic to create the initial VPlan to a separate helper, and only instantiate `VPRecipeBuilder` there

**Comment by ayalz** [comment]
> > Independent of this patch: should VPRecipeBuilder (or some other sub/class thereof) be confined to initial VPlan construction from LLVM-IR, excluding construction of recipes during VPlan-to-VPlan transformations?
> 
> 
> 
> Yes, `VPRecipeBuilder` ( not to be confused with `VPBuilder`) is only used for initial construction. We should split of the logic to create the initial VPlan to a separate helper, and only instantiate `VPRecipeBuilder` there

Agreed. Would also be good to clarify the distinction between the two builders, possibly using more accurate names. May also be interesting to consider the analogous TransformState, used to maintain the temporary state during translation from VPlan to LLVM-IR.

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Certainly more accurate (LiveIn is also a VPValue) and compact. Worth pushing independently.

**Review by ayalz — APPROVED** [review]
LGTM, a comment by @nikolaypanchenko awaits response.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 7]
```diff
@@ -7938,7 +7950,7 @@ VPValue *VPRecipeBuilder::createEdgeMask(BasicBlock *Src, BasicBlock *Dst,
   if (OrigLoop->isLoopExiting(Src))
     return EdgeMaskCache[Edge] = SrcMask;
 
-  VPValue *EdgeMask = Plan.getVPValueOrAddLiveIn(BI->getCondition());
+  VPValue *EdgeMask = getVPValueOrAddLiveIn(BI->getCondition(), Plan);
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7937]: Forgot to mention, this has been split off & landed in 8578b6e9120cb2fc8a7fc102587d617144c35532, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 8]
```diff
@@ -54,7 +54,6 @@ void VPlanTransforms::VPInstructionsToVPRecipes(
               vputils::getOrCreateVPValueForSCEVExpr(*Plan, II->getStep(), SE);
           NewRecipe = new VPWidenIntOrFpInductionRecipe(Phi, Start, Step, *II);
         } else {
-          Plan->addVPValue(Phi, VPPhi);
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:57]: Done thanks!

**Pull request merged**